### PR TITLE
kn: checksum bindings

### PR DESCRIPTION
**Pull request by lauzadis**
*Description of changes:*

- Add MD5 implementation
- Add `WithCrt` mixin class used to ensure `CRT.initRuntime { }` is called
    - Trying to use these CRT hash functions from smithy-kotlin would result in errors because CRT was not properly initialized
- Fix an `IndexOutOfBounds` bug caused by passing empty inputs to hash functions    

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+package aws.sdk.kotlin.crt
+
+/**
+ * A mixin class used to ensure CRT is initialized before the class is invoked
+ */
+public open class WithCrt {
+    init {
+        CRT.initRuntime { }
+    }
+}
```
Comment (ianbotsf) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt:14]: **Question**: I'm concerned with this approach for two reasons:
1. Kotlin doesn't support true mixin classes so this pattern prevents inheritors from using any other superclass. Given that this functionality isn't _truly_ polymorphic it seems prudent to leave the option open.
1. This initializes the CRT runtime at object instantiation rather than when CRT functionality is actually necessary. If some code just instantiates a new `Crc32` but never uses any methods on it, we'll've wasted cycles unnecessarily.

Could we move the initialization closer to the site of use?

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/CrcNative.kt [hunk 2]
```diff
@@ -21,6 +24,10 @@ internal class Crc(val checksumFn: AwsChecksumsCrcFunction) : HashFunction {
     private var crc = 0U
 
     override fun update(input: ByteArray, offset: Int, length: Int) {
+        if (input.isEmpty() || length == 0) {
+            return
+        }
+
```
Comment (ianbotsf) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/CrcNative.kt:30]: **Question**: Why is this necessary? What breaks if we just pass an empty blob or a 0 length down to CRT?

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt (hunk 1 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt:14]: > Could we move the initialization closer to the site of use?

Sure, we'll just have to take care to initialize CRT everywhere we use it, which we currently don't do.

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/CrcNative.kt (hunk 2 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/CrcNative.kt:30]: The request would never make it to CRT, the following code snippet fails with an `ArrayIndexOutOfBoundsException` exception:

```
        val offsetInput = input.usePinned {
            it.addressOf(offset)
        }
```

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt (hunk 1 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt:14]: > If some code just instantiates a new Crc32 but never uses any methods on it, we'll've wasted cycles unnecessarily.

This is true but in practice CRT is only initialized once, subsequent calls to `CRT.initRuntime { }` are no-ops. I'm assuming users of aws-sdk-kotlin Kotlin/Native will need to initialize CRT at some point

Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/WithCrt.kt:14]: We spoke offline and agreed to continue this approach for now. Ideally we would create a singleton object `CrtLib` and use cinterop to generate extensions for it (like `CrtLib.aws_checksums_crc32(...)`). This `CrtLib` object would ensure `CRT.initRuntime { ... }` is called before invoking the actual C function.
